online-shop/src/app/cart/state/cart-reducers.ts

```diff
@@ -0,0 +1,60 @@
+import { Product } from 'src/app/classes/product';
+import { createReducer, on } from '@ngrx/store';
+import {
+  addToCart,
+  checkout,
+  checkoutFailure,
+  checkoutSuccess,
+  clearCart,
+} from './cart-actions';
+
+export interface CartState {
+  products: Product[];
+  status: 'error' | 'pending' | 'success' | 'processing';
+  error: String | null;
+}
+
+export const initialCartState: CartState = {
+  products: [],
+  status: 'success',
+  error: null,
+};
+
+export const cartReducer = createReducer(
+  initialCartState,
+  on(checkout, (state) => ({ ...state, status: 'processing' })),
+  on(clearCart, (state) => ({ ...state, status: 'success', products: [] })),
+  on(checkoutSuccess, (state) => ({
+    ...state,
+    status: 'success',
+    products: [],
+  })),
+  //The http response that is returned cannot be parsed as json, which always gives an error. That's why the error and success order creation actions are the same.
+  on(checkoutFailure, (state, { error }) => ({
+    ...state,
+    status: 'success',
+    products: [],
+  })),
+  on(addToCart, (state, { product }) => ({
+    ...state,
+    status: 'pending',
+    products: ((products: Product[], addedProduct: Product) => {
+      let newProducts = [];
+      let exists = false;
+      for (let product of products) {
```

Comment:
I think this logic is a bit more complicated than it should. as I see it, it's:
found = products.find(product => product.id === addedProduct.id && product.quantity !== undefined);
if(found) found.quantity++
else products.push(newProduct, quantity: 1)
found will be another reference pointing to the object in the array, altering it will really alter the value inside your array.

Reply:
Actually, all the object I use inside the arrow function seem to be read only, so each product must be individually copied/modified and added to another list.